cumulusci/robotframework/form_handlers.py

```diff
@@ -0,0 +1,162 @@
+from robot.libraries.BuiltIn import BuiltIn
+from robot.api import logger
+from selenium.webdriver.common.action_chains import ActionChains
+from cumulusci.utils import get_all_subclasses
+
+"""
+FIXME: Something that still needs to be resolved is what to do about
+name clashes. What if a project imports two different libraries and
+each library implements a handler for the same tag and/or a handler
+for which a built-in handler already exists. How do we decide what
+takes precedence. Is this even something we have to worry about?
+
+So, the idea is to find an element, then do something like
+BaseFormHandler.get_handler(element, locator).set("foo")
+
+"""
+
+
+def get_form_handler(element, locator):
+    """Return an instance of handler for the given element
+
+    This will search all subclasses of BaseFormHandler, looking
+    for a class that supports the tag_name of the given element
+    """
+    tag = element.tag_name
+    for subclass in get_all_subclasses(BaseFormHandler):
+        if tag in subclass.tags:
+            # first one wins, but is that the right algorithm? Do
+            # we want to give precedence to external handlers
+            # over our built-in ones?
+            return subclass(element, locator)
+    return None
+
+
+class BaseFormHandler:
+    def __init__(self, element, locator):
+        # concrete element to be handled
+        self.element = element
+
+        # the locator used to get the element, which we use for error
+        # handling.
+        self.locator = locator
+
+    @property
+    def selenium(self):
+        return BuiltIn().get_library_instance("SeleniumLibrary")
+
+    @property
+    def input_element(self):
+        """Returns the actual <input> or <textarea> element inside the element"""
+        elements = self.element.find_elements_by_xpath(
+            ".//*[self::input or self::textarea]"
+        )
+        return elements[0] if elements else None
+
+    # these should be implemented by each of the handlers
```

Comment:
It would be best to make BaseFormHandler a subclass of abc.ABC, and mark the abstract methods using the abstractmethod decorator. That way Python will raise an error if someone tries to implement a subclass without providing a concrete implementation of one of the abstract methods. Here's an example from David Reed's dependency refactoring: https://github.com/SFDO-Tooling/CumulusCI/blob/3b4a4d9d538c93707e2984b3a4653b8bd322e4ae/cumulusci/core/dependencies/dependencies.py#L67

Another benefit is that we have coverage.py configured to skip requiring coverage for abstract methods.

Reply:
Are focus and clear required or not? Not all of the handlers implement them, but I guess the keywords aren't making use of them either at this point.

Reply:
Eventually, I think we'll require them, but they aren't required at the moment. I made `set`, `get`, and `clear` abstract methods to show where the code is heading. `focus` is the one that probably won't need to be implemented for each handler.